src/protocol/boolean/prefix_or.rs

```diff
@@ -0,0 +1,365 @@
+    ///
+    /// # Example
+    /// ```ignore
+    ///   [f] = 0 1 0 0
+    ///   [b] = 0 0 1 1
+    ///   [s] = 0 0 0 0  0 0 1 1  0 0 0 0  0 0 0 0
+    /// ```
+    async fn step8(
+        f: &[Replicated<B>],
+        b: &[Replicated<B>],
+        ctx: ProtocolContext<'_, B>,
+    ) -> Result<Vec<Replicated<B>>, BoxError> {
+        let lambda = f.len();
+        let mut mul = Vec::new();
+        for (i, &f_bit) in f.iter().enumerate().take(lambda) {
+            for (j, &b_bit) in b.iter().enumerate().take(lambda) {
+                let c = ctx.clone();
+                mul.push(
+                    c.multiply(RecordId::from(lambda * i + j))
+                        .execute(f_bit, b_bit),
+                );
+            }
+        }
+        try_join_all(mul).await
+    }
+
+    /// Step 9. `for i,j=1..λ, [b_(i,j)] = [s_(i,j)] + [y_i] - [f_i]`
+    ///
+    /// # Example
+    /// ```ignore
+    ///   [s] = 0 0 0 0  0 0 1 1  0 0 0 0  0 0 0 0
+    ///   [y] = 0 1 1 1
+    ///   [f] = 0 1 0 0
+    ///   [b] = 0 0 0 0  0 0 1 1  1 1 1 1  1 1 1 1   // <- PrefixOr([a])
+    /// ```
+    fn step9(s: &[Replicated<B>], y: &[Replicated<B>], f: &[Replicated<B>]) -> Vec<Replicated<B>> {
+        let lambda = f.len();
+        (0..lambda)
+            .flat_map(|i| {
+                (0..lambda)
+                    .map(|j| s[lambda * i + j] + y[i] - f[i])
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// Execute `PrefixOr`
+    #[allow(dead_code)]
+    #[allow(clippy::many_single_char_names)]
+    pub async fn execute(
+        &self,
+        ctx: ProtocolContext<'_, B>,
+    ) -> Result<Vec<Replicated<B>>, BoxError> {
+        // The paper assumes `l = λ^2`, where `l` is the bit length of the input
+        // share. Then the input is split into `λ` blocks each holding `λ` bits.
+        // Or operations are executed in parallel by running the blocks in
+        // parallel.
+        //
+        // The paper doesn't mention about cases where `l != λ^2`. For now, we
+        // pad the input with dummy bits at the end, and will be stripped off
+        // before returning the output to the caller. The output is correct
+        // regardless of the dummy bits, but may affect the performance if
+        // `λ^2 - l` becomes large. We should revisit this protocol once the
+        // prototype is complete, and optimize if necessary.
+        // TODO(taikiy): Prefix-Or optimization
+
+        // It should be safe to cast from `usize` to `u32` and back cast since
+        // the input we expect to operate on are 40-bit < `u8` << `u32`.
+        #[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
+        let lambda = f64::from(self.input.len() as u32).sqrt().ceil() as usize;
+        let dummy = vec![Replicated::new(B::ZERO, B::ZERO); lambda * lambda - self.input.len()];
```

Comment:
I guess we need this as we plan to utilize u32... so we will pad out 4 dummy values to get to 36?

But we don't really want to compute a square root each time we invoke this protocol do we? So maybe we should special case it? like match on input length and fatal if it's not coded for?

Reply:
I changed to code to assert if the input length is not either 8, 16, or 32. If the length is one of these, I return 3, 4 or 6 as lambda. Anything else, the protocol panics.